### PR TITLE
[shelly] Add Plus/Pro support, some bugfixes

### PR DESCRIPTION
**Pull request by markus7017**
This PR replaces #12842

The PR adds support for the Shelly Generation 2 Plus/Pro Devices .

I tried to separate it from other changes (previous PRs), but it still has a relevant size.

### CODE REVIEW

**Comment by jlaur** [comment]
@markus7017 - can you close the PR being replaced?

**Comment by markus7017** [comment]
@fwolter @lolodomo I fixed the merge conflict and ready to start the review. Pinged you both, because you were in previous PRs.

**Review by fwolter — CHANGES_REQUESTED** [review]
Can you say a few words about how Gen2 devices connect in general? Do the devices connect to the OH instance? Is there some authentification? Is it save to expose the servlet to the internet?

**Comment by markus7017** [comment]
> Can you say a few words about how Gen2 devices connect in general? Do the devices connect to the OH instance? Is there some authentification? Is it save to expose the servlet to the internet?

The binding supports Generation 1 (first series) and Plus / Pro series of devices, which are Generation 2. Both generations uses complete different APIs
- Generation 1: Uses http-based communication with basic authentication. In addition the devices implement CoIoT version 1 + 2, a Allterco specific IoT protocol based on Coap. Therefore the Californium framework is integrated to handle the CoAP protocol. CoAP supports multicast mode and peer mode. Using the peer mode is an easy way to have specific firewall rules protecting the device access.
- Generation 2: Complete different API, which is based on http POSTs for device control and an WebSocket interface for event messages. In addition the binding implements an inbound WebSocket, which is required to receive wakeup / status notifications from battery powered devices, which go into sleep mode. Gen2 uses digest authentication if restricted access has been enabled in the device setup. The user is "admin" (hard-code for whatever reason) and password is taken from the binding or thing configuration.  The device also supports secure WebSocket communication (wss, incl. X.5409 certificates), but this is not yet implemented in the binding. In general you might access a Shelly device directly from the Internet or exposing the servlet entry point using protected device setup. However, I would prefer a VPN in between.
- ShellyManager is also based on a Servlet and implements some advanced device diagnostic and control functions. The Manager runs in the openHAB context so you need to login as usual being able to access the UI (http / https).

Short note on the API. Due to the fact that APIs are different, also the data formats, but device functionality is similar I decided to implement a mapping layer. Data structures are mapped form Gen2 format to Gen1 format for further processing on the upper code layers. Otherwise I would have to re-implement the complete logic, in fact redundant. Gen1 devices are still in production so removing Gen1 support was no option. Also creating a 2nd binding would increase maintenance efforts and may confuses the users.
I separate Gen1 and Gen2 APIs in packages
- api1 = Gen1-specific 
- api2 = Gen2-spcific
- apu = Common code
- ShellyApiInterface defines the API interface, which then has 2 implementations

Please feel free to ask further questions if this gets a better understanding for the review.

File: bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/thing/shellyGen2_relay.xml [hunk 1]
```diff
@@ -0,0 +1,259 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<thing:thing-descriptions bindingId="shelly"
+	xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+	xmlns:thing="https://openhab.org/schemas/thing-description/v1.0.0"
+	xsi:schemaLocation="https://openhab.org/schemas/thing-description/v1.0.0 https://openhab.org/schemas/thing-description-1.0.0.xsd">
+
+	<thing-type id="shellyplus1">
+		<label>ShellyPlus 1 (SNSW-001X16EU")</label>
```
[the remark this answers is not in the document]
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/resources/OH-INF/thing/shellyGen2_relay.xml:8]: Technically yes, but from a user's point of view no, so I removed all incl. those for Gen 1 devices

**Comment by markus7017** [comment]
@fwolter requested changes are resolved

File: bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api2/Shelly2RpcSocket.java [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+                            session.getRemoteAddress());
+                    s.disconnect();
+                }
+                s.close(StatusCode.NORMAL, "Socket closed");
+                session = null;
+            }
+            if (client.isStarted()) {
+                client.stop();
+            }
+        } catch (/* IOException | */Exception e) {
```
[the remark this answers is not in the document]
Comment (fwolter) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api2/Shelly2RpcSocket.java:215]: Can you comment on this?

[the remark this answers is not in the document]
Comment (markus7017) [reply bundles/org.openhab.binding.shelly/src/main/java/org/openhab/binding/shelly/internal/api2/Shelly2RpcSocket.java:215]: removed (not yet checked in)

**Comment by markus7017** [comment]
@fwolter I pushed the changes, also removing some other left overs from development phase :-)

**Review by fwolter — APPROVED** [review]
Sounds good!

**Pull request merged**